sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java

```diff
@@ -539,16 +539,14 @@ public static DynamicMessage messageFromMap(
 
         if (fieldSchemaInformation.getType().equals(TableFieldSchema.Type.STRUCT)
             && unknownFields != null) {
-          if (unknownFields.get(key) instanceof Map
-              && ((Map<?, ?>) unknownFields.get(key)).isEmpty()) {
+          if ((unknownFields.get(key) instanceof Map
+                  && ((Map<?, ?>) unknownFields.get(key)).isEmpty()) // single struct, empty
+              || (unknownFields.get(key)
+                      instanceof List // repeated struct, empty list or list with empty structs
+                  && (((List<?>) unknownFields.get(key)).isEmpty()
+                      || ((List<?>) unknownFields.get(key))
+                          .stream().allMatch(row -> row == null || ((Map<?, ?>) row).isEmpty())))) {
```

Comment:
I guess we can also merge with the outer if-statement? this is gonna be a monster of an if-statement.

Maybe there's a way to clean it up, but it's not PR-blocking so will leave it up to you.

Reply:
Done and green